### PR TITLE
Audio provider

### PR DESCRIPTION
**Pull request by szymon-filipiak**
Merge changes into the voice assistant branch. @plonajakub  For the purpose of keeping code review alive please do not merge other branches into your working branch, as it is difficult to review changes that way. Let's check this code now, and we will resolve who should fix problems, so @zawadzki-dawid  be informed.

### CODE REVIEW

File: tensorflow/lite/micro/examples/micro_speech/Makefile.inc [hunk 1]
```diff
@@ -5,54 +5,71 @@ INCLUDES += \
 GENERATED_PROJECT_INCLUDES += \
 -I./third_party/kissfft
 
+#DONE FROM HERE
```
Comment (szymon-filipiak) [inline tensorflow/lite/micro/examples/micro_speech/Makefile.inc]: Are those comments necessary? Can we remove them?

File: tensorflow/lite/micro/examples/micro_speech/Makefile.inc [hunk 2]
```diff
@@ -89,65 +108,87 @@ tensorflow/lite/experimental/microfrontend/lib/pcan_gain_control_util.h \
...
 MICRO_FEATURES_GENERATOR_TEST_SRCS := \
 tensorflow/lite/micro/examples/micro_speech/micro_features/micro_features_generator_test.cc \
 tensorflow/lite/micro/examples/micro_speech/micro_features/no_feature_data_slice.cc \
 tensorflow/lite/micro/examples/micro_speech/micro_features/yes_feature_data_slice.cc \
 tensorflow/lite/micro/examples/micro_speech/no_30ms_sample_data.cc \
 tensorflow/lite/micro/examples/micro_speech/yes_30ms_sample_data.cc \
 $(MICRO_FEATURES_GENERATOR_SRCS)
+#TO HERE
```
Comment (szymon-filipiak) [inline tensorflow/lite/micro/examples/micro_speech/Makefile.inc]: A lot of comments :P

File: tensorflow/lite/micro/examples/micro_speech/main_functions.cc [hunk 3]
```diff
@@ -12,12 +12,10 @@ WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 See the License for the specific language governing permissions and
 limitations under the License.
 ==============================================================================*/
-
 #include "tensorflow/lite/micro/examples/micro_speech/main_functions.h"
-
+#include "tensorflow/lite/micro/examples/micro_speech/feature_provider.h"
```
Comment (szymon-filipiak) [inline tensorflow/lite/micro/examples/micro_speech/main_functions.cc]: Is this change necessary? I think that earlier all includes were sorted more or less alphabetically except for main_functions.h which is heder of this file

File: tensorflow/lite/micro/examples/micro_speech/zephyr_riscv/CMakeLists.txt [hunk 4]
```diff
@@ -0,0 +1,65 @@
+cmake_minimum_required(VERSION 3.13.1)
+find_package(Zephyr)
```
Comment (szymon-filipiak) [inline tensorflow/lite/micro/examples/micro_speech/zephyr_riscv/CMakeLists.txt]: I suggest setting the `REQUIRED`  option in `find_package`. It stops compilation if the package is not found. More information [here](https://cmake.org/cmake/help/latest/command/find_package.html). What do you think?

File: tensorflow/lite/micro/examples/micro_speech/zephyr_riscv/CMakeLists.txt [hunk 5]
```diff
@@ -0,0 +1,65 @@
...
+project(tf_lite_micro_speech)
+
+set(TF_MICRO_DIR ${PROJECT_SOURCE_DIR}/../../../)
+file(GLOB MICRO_SOURCES ${TF_MICRO_DIR}/*cc)
```
Comment (szymon-filipiak) [inline tensorflow/lite/micro/examples/micro_speech/zephyr_riscv/CMakeLists.txt:6]: Globing files in CMake is not a recommended way of attaching sources to the project. It's the easiest approach but comes with a cost of **not** reloading CMake if files are changed. To make sure it's working correctly you must reload CMake after each file modification (or git change) in that directory (or modify CMakeLists.txt, then CMake will reload automatically).

It's not a big deal to remember that, but we must be aware of it.

Another way to include those files is to list them explicitly. Then adding/deleting a file requires modifying CMakeLists, but CMake will take care of tracking file content changes. I invite you to the discussion. Check out [this](https://cmake.org/cmake/help/latest/command/file.html#glob) and [this](https://stackoverflow.com/questions/1027247/is-it-better-to-specify-source-files-with-glob-or-each-file-individually-in-cmak/1060061#1060061)

File: tensorflow/lite/micro/examples/micro_speech/zephyr_riscv/CMakeLists.txt [hunk 6]
```diff
@@ -0,0 +1,65 @@
...
+# YOU CAN FIND SUBDIRECTORIES IN cmake_sub/ directory
+add_subdirectory(cmake_sub/micro_features)
+add_subdirectory(cmake_sub/simple_features)
+add_subdirectory(cmake_sub/micro_speech)
+add_subdirectory(cmake_sub/kiss_fft)
+add_subdirectory(cmake_sub/microfrontend)
+
+target_sources(
+	app PRIVATE
+        src/audio_provider.cc
```
Comment (szymon-filipiak) [inline tensorflow/lite/micro/examples/micro_speech/zephyr_riscv/CMakeLists.txt]: Can we make a lib from this sources as discussed on the meeting?

File: tensorflow/lite/micro/examples/micro_speech/zephyr_riscv/CMakeLists.txt [hunk 7]
```diff
@@ -0,0 +1,65 @@
...
+	${TF_MICRO_DIR}/../../..
+	${TF_MICRO_DIR}/tools/make/downloads/ruy
+        ${TF_MICRO_DIR}/../micro/tools/make/downloads/gemmlowp
+        ${TF_MICRO_DIR}/../micro/tools/make/downloads/flatbuffers/include
+        $ENV{TOOLCHAIN_BASE}/include/c++/$ENV{TOOLCHAIN_VERSION}
+        $ENV{TOOLCHAIN_BASE}/include
+        $ENV{TOOLCHAIN_BASE}/include/c++/$ENV{TOOLCHAIN_VERSION}/riscv32-zephyr-elf/rv32i/ilp32
+        )
+
+add_definitions(-DNDEBUG -DTF_LITE_STATIC_MEMORY)
```
Comment (szymon-filipiak) [inline tensorflow/lite/micro/examples/micro_speech/zephyr_riscv/CMakeLists.txt]: CMake documentation says that this command has been superseded by alternatives. See [this](https://cmake.org/cmake/help/v3.13/command/add_definitions.html). I think we should use `add_compile_options()` here, but I'm not sure.

File: tensorflow/lite/micro/examples/micro_speech/zephyr_riscv/Makefile.inc [hunk 8]
```diff
@@ -0,0 +1,6 @@
+ifeq ($(TARGET), zephyr_vexriscv)
+	export ZEPHYR_TOOLCHAIN_VARIANT?=zephyr
+micro_speech_bin:
+	cmake -B ${GENDIR}/micro_speech/CMake -GNinja -DBOARD="litex_vexriscv" -S tensorflow/lite/micro/examples/micro_speech/zephyr_riscv
```
Comment (szymon-filipiak) [inline tensorflow/lite/micro/examples/micro_speech/zephyr_riscv/Makefile.inc]: Maby we can consider passing -Wdev option by default, to enable warnings?

File: tensorflow/lite/micro/examples/micro_speech/zephyr_riscv/cmake_sub/kiss_fft/CMakeLists.txt [hunk 9]
```diff
@@ -0,0 +1,14 @@
+# /tensorflow/tensorflow/lite/micro/tools/make/downloads/kissfft
+add_library (
+	kiss_fft
+	../../../../../tools/make/downloads/kissfft/kiss_fft.c
```
Comment (szymon-filipiak) [inline tensorflow/lite/micro/examples/micro_speech/zephyr_riscv/cmake_sub/kiss_fft/CMakeLists.txt]: Please extract paths to directory as discussed in the meeting.

File: tensorflow/lite/micro/examples/micro_speech/zephyr_riscv/cmake_sub/micro_features/CMakeLists.txt [hunk 10]
```diff
@@ -0,0 +1,25 @@
+# /tensorflow/tensorflow/lite/micro/examples/micro_speech/micro_features
+add_library (
+	micro_features
+	../../../micro_features/tiny_conv_micro_features_model_data.cc
+	../../../micro_features/tiny_conv_micro_features_model_data.h
```
Comment (szymon-filipiak) [inline tensorflow/lite/micro/examples/micro_speech/zephyr_riscv/cmake_sub/micro_features/CMakeLists.txt]: Same as for kiss_fft

File: tensorflow/lite/micro/examples/micro_speech/zephyr_riscv/prj.conf [hunk 11]
```diff
@@ -0,0 +1,2 @@
+CONFIG_CPLUSPLUS=y
+CONFIG_NEWLIB_LIBC=y
```
Comment (szymon-filipiak) [inline tensorflow/lite/micro/examples/micro_speech/zephyr_riscv/prj.conf:2]: What are those lines doing? Are those Zephyr constants?

File: tensorflow/lite/micro/tools/make/download_and_extract.sh [hunk 12]
```diff
@@ -86,17 +86,6 @@ build_embarc_mli() {
...
-  command -v virtualenv >/dev/null 2>&1 || {
-    echo >&2 "The required 'virtualenv' tool isn't installed. Try 'pip install virtualenv'."; exit 1;
-  }
-  virtualenv -p python3 ${1}/venv-zephyr
-  . ${1}/venv-zephyr/bin/activate
-  python ${1}/venv-zephyr/bin/pip install -r ${1}/scripts/requirements.txt
-  west init -m https://github.com/zephyrproject-rtos/zephyr.git
-  deactivate
-}
-
```
Comment (szymon-filipiak) [inline tensorflow/lite/micro/tools/make/download_and_extract.sh:99]: Should we delete this? Maby rather adapt to clone our repo or something like this?

**Review by szymon-filipiak — COMMENTED** [review]
I need to do it differently because the author of the merge request can't properly review the code. Please @plonajakub @zawadzki-dawid  submit feedback to the comments I've created and resolve all threads.

File: tensorflow/lite/micro/examples/micro_speech/Makefile.inc (hunk 1 above)
Comment (zawadzki-dawid) [reply tensorflow/lite/micro/examples/micro_speech/Makefile.inc]: Nope, i did them for my work convenience, i will delete them.

File: tensorflow/lite/micro/examples/micro_speech/zephyr_riscv/CMakeLists.txt (hunk 4 above)
Comment (zawadzki-dawid) [reply tensorflow/lite/micro/examples/micro_speech/zephyr_riscv/CMakeLists.txt]: I have tested out compilation with and without `REQUIRED` option. Compilation with 
`REQUIRED` throws an error much faster, i think it's a good idea to add it.

File: tensorflow/lite/micro/examples/micro_speech/main_functions.cc (hunk 3 above)
Comment (zawadzki-dawid) [reply tensorflow/lite/micro/examples/micro_speech/main_functions.cc]: It's not necessary, just a random change while fixing a bug. We can revert it to previous state.

File: tensorflow/lite/micro/examples/micro_speech/zephyr_riscv/CMakeLists.txt (hunk 5 above)
Comment (Marzecki) [reply tensorflow/lite/micro/examples/micro_speech/zephyr_riscv/CMakeLists.txt:6]: The thing is I just copied the whole CMakeList.txt from the magic_wand example and adapted it so that it would contain the right files needed for us. Didn't really think about the globing but I agree with you.

File: tensorflow/lite/micro/examples/micro_speech/zephyr_riscv/CMakeLists.txt [hunk 13]
```diff
@@ -0,0 +1,72 @@
...
+		../micro_features/tiny_conv_micro_features_model_data.cc
+		../micro_features/tiny_conv_micro_features_model_data.h
+
+                #../output_handler.cc
+                #../output_handler.h
+                #../gesture_predictor.cc
+                #../constants.h
+                #../gesture_predictor.h
+                #../magic_wand_model_data.cc
+		#../magic_wand_model_data.h
```
Comment (Marzecki) [inline tensorflow/lite/micro/examples/micro_speech/zephyr_riscv/CMakeLists.txt]: These comments can be deleted.

File: tensorflow/lite/micro/examples/micro_speech/zephyr_riscv/CMakeLists.txt (hunk 7 above)
Comment (Marzecki) [reply tensorflow/lite/micro/examples/micro_speech/zephyr_riscv/CMakeLists.txt]: As with the globing. We can change it and find out if it still works.

File: tensorflow/lite/micro/examples/micro_speech/zephyr_riscv/prj.conf (hunk 11 above)
Comment (Marzecki) [reply tensorflow/lite/micro/examples/micro_speech/zephyr_riscv/prj.conf:2]: Same as globing. These are features provided by Zephyr’s Kconfig configuration system. You can see [this](https://docs.zephyrproject.org/1.13.0/application/application.html#creating-an-application) (point 7). I guess they enable using C/C++ files.

File: tensorflow/lite/micro/examples/micro_speech/zephyr_riscv/CMakeLists.txt (hunk 5 above)
Comment (zawadzki-dawid) [reply tensorflow/lite/micro/examples/micro_speech/zephyr_riscv/CMakeLists.txt:6]: It's good to know that, but is that a big problem in our project? I think we can leave it as it is, code is more readable and according to docs you linked it's not unstable in any way, just have to know the cons. Are we going to modify those files anyway?

File: tensorflow/lite/micro/examples/micro_speech/zephyr_riscv/CMakeLists.txt (hunk 6 above)
Comment (zawadzki-dawid) [reply tensorflow/lite/micro/examples/micro_speech/zephyr_riscv/CMakeLists.txt]: Of course sir <3

File: tensorflow/lite/micro/examples/micro_speech/zephyr_riscv/prj.conf (hunk 11 above)
Comment (szymon-filipiak) [reply tensorflow/lite/micro/examples/micro_speech/zephyr_riscv/prj.conf:2]: Okay, thanks!

File: tensorflow/lite/micro/examples/micro_speech/zephyr_riscv/CMakeLists.txt (hunk 5 above)
Comment (szymon-filipiak) [reply tensorflow/lite/micro/examples/micro_speech/zephyr_riscv/CMakeLists.txt:6]: I'm just pointing this out now, to highlight possible error prone areas. I don't know if this will be a problem in our case. I think we should leave this as it is for now and refactor in the future if necessary.

File: tensorflow/lite/micro/examples/micro_speech/zephyr_riscv/CMakeLists.txt (hunk 7 above)
Comment (zawadzki-dawid) [reply tensorflow/lite/micro/examples/micro_speech/zephyr_riscv/CMakeLists.txt]: `add_compile_options` works fine. I have superseded `add_definitions` with this one.

File: tensorflow/lite/micro/examples/micro_speech/zephyr_riscv/Makefile.inc (hunk 8 above)
Comment (zawadzki-dawid) [reply tensorflow/lite/micro/examples/micro_speech/zephyr_riscv/Makefile.inc]: Sounds good to me, we can do that.

**Comment by zawadzki-dawid** [comment]
@szymon-filipiak I have pushed my changes with all the remarks you have. Looks like we can merge this branch.

**Comment by szymon-filipiak** [comment]
So @plonajakub can you merge David's branch locally and push it here? Just to avoid any further conflicts. Then I will resolve this pull request if there is no more comments. This is the first merge request and it came out a little bit messy, but I would like to discuss the process at the next meeting :D

**Comment by szymon-filipiak** [comment]
I've talked with Jakub and he gave me permit to merge changes to his branch.

File: tensorflow/lite/micro/examples/micro_speech/zephyr_riscv/Makefile.inc (hunk 8 above)
Comment (plonajakub) [reply tensorflow/lite/micro/examples/micro_speech/zephyr_riscv/Makefile.inc]: So as I can read in the CMake docs the option does as follows.

> Enable developer warnings.
>Enable warnings that are meant for the author of the CMakeLists.txt files. By default, this will also turn on deprecation warnings.

I have nothing against doing this.

File: tensorflow/lite/micro/tools/make/download_and_extract.sh (hunk 12 above)
Comment (plonajakub) [reply tensorflow/lite/micro/tools/make/download_and_extract.sh:99]: I have deleted this because I assumed every one of us will do the first steps from zephyr getting started guide. If so there is no need to keep these lines and I've done that to keep the codebase nice&clean.

File: tensorflow/lite/micro/examples/micro_speech/zephyr_riscv/CMakeLists.txt (hunk 13 above)
Comment (plonajakub) [reply tensorflow/lite/micro/examples/micro_speech/zephyr_riscv/CMakeLists.txt]: It is an old version of CMake so the lines are no longer the case.

File: tensorflow/lite/micro/examples/micro_speech/zephyr_riscv/Makefile.inc (hunk 8 above)
Comment (plonajakub) [reply tensorflow/lite/micro/examples/micro_speech/zephyr_riscv/Makefile.inc]: Expect an update in an hour.

**Comment by szymon-filipiak** [comment]
It looks good.
